src/lint/rules/algorithm-line-endings.ts

```diff
@@ -0,0 +1,203 @@
+      }
+      let firstIndex = 0;
+      while (firstIndex < node.contents.length && node.contents[firstIndex].name === 'tag') {
+        ++firstIndex;
+      }
+      if (firstIndex === node.contents.length) {
+        report({
+          line: node.location!.start.line,
+          column: node.location!.start.column,
+          message: 'expected line to contain non-tag elements',
+        });
+        return;
+      }
+      let first = node.contents[firstIndex];
+
+      let last = node.contents[node.contents.length - 1];
+
+      // Special case: if the step has a figure, it should end in `:`
+      if (last.name === 'tag' && last.contents === '</figure>') {
+        let count = 1;
+        let lastIndex = node.contents.length - 2;
+        if (lastIndex < 0) {
+          report({
+            line: node.location!.start.line,
+            column: node.location!.start.column,
+            message: 'could not find matching <figure> tag',
+          });
+          return;
+        }
+        while (count > 0) {
+          last = node.contents[lastIndex];
+          if (last.name === 'tag') {
+            if (last.contents === '<figure>') {
+              --count;
+            } else if (last.contents === '</figure>') {
+              ++count;
+            }
+          }
+          --lastIndex;
+          if (lastIndex < 0) {
+            report({
+              line: node.location!.start.line,
+              column: node.location!.start.column,
+              message: 'could not find matching <figure> tag',
+            });
+            return;
+          }
+        }
+        last = node.contents[lastIndex];
+        if (last.name !== 'text') {
+          report({
+            line: node.location!.start.line,
+            column: node.location!.start.column,
+            message: `expected line to end with text (found ${last.name})`,
+          });
+          return;
+        }
+        if (!/:\n +$/.test(last.contents)) {
+          report({
+            line: node.location!.start.line,
+            column: node.location!.start.column,
+            message: 'expected line with figure to end with ":"',
+          });
+        }
+        return;
+      }
+
+      if (last.name !== 'text') {
+        report({
+          line: node.location!.start.line,
+          column: node.location!.start.column,
+          message: `expected line to end with text (found ${last.name})`,
+        });
+        return;
+      }
+
+      let initialText = first.name === 'text' ? first.contents : '';
+      let hasSubsteps = node.sublist !== null;
+
+      if (/^(?:If |Else if)/.test(initialText)) {
+        if (hasSubsteps) {
+          if (node.sublist!.name === 'ol') {
+            if (!/, then$/.test(last.contents)) {
+              report({
+                line: node.location!.start.line,
+                column: node.location!.start.column,
+                message: `expected "If" with substeps to end with ", then" (found "${last.contents}")`,
```

Comment:
I kinda prefer quoting these things with `JSON.stringify` to avoid surprising messages.

Reply:
or https://npmjs.com/object-inspect, if the values might not be JSON-serializable

Reply:
> I kinda prefer quoting these things with JSON.stringify to avoid surprising messages.

Done.

> if the values might not be JSON-serializable

They're all strings.